utils/buffer/README.md

```diff
@@ -1,105 +1,151 @@
+- `get_data_owned(..)` returns the inner vector.
 
-## BufferPool
-Usually `BufferFromSystemMemory` should be enough, but sometimes it is better to use something faster.
+## `BufferPool`
+While `BufferFromSystemMemory` is sufficient for many cases, `BufferPool` offers a more efficient
+solution for high-performance applications, such as proxies and pools with thousands of connections.
 
-For each Sv2 connection, there is a `Decoder` and for each decoder, there are 1 or 2 buffers.
+When created, `BufferPool` preallocates a user-defined capacity of bytes in the heap using a
+`Vec<u8>`. When `get_data_owned(..)` is called, it creates a `Slice` that contains a view into the
+preallocated memory. `BufferPool` guarantees that slices never overlap and maintains unique
+ownership of each `Slice`.
 
-Proxies and pools with thousands of connections should use `Decoder<BufferPool>` rather than
-`Decoder<BufferFromSystemMemory>`
+`Slice` implements the `Drop`, allowing the view into the preallocated memory to be reused upon
+dropping.
 
-`BufferPool` when created preallocate a user-defined capacity of bytes in the heap using a
-`Vec<u8>`, then when `get_data_owned(..)` is called it create a `Slice` that contains a view into
-the preallocated memory. `BufferPool` guarantees that slices never overlap and the uniqueness of
-the `Slice` ownership.
+### Buffer Management and Allocation
 
-`Slice` implements `Drop` so that the view into the preallocated memory can be reused.
+`BufferPool` is useful for working with sequentially processed buffers, such as filling a buffer,
+retrieving it, and then reusing it as needed. `BufferPool` optimizes for memory reuse by providing
+pre-allocated memory that can be used in one of three modes:
 
-### Fragmentation overflow and optimization
-`BufferPool` can allocate a maximum of 8 `Slices` (cause it uses an `AtomicU8` to keep track of the
-used and freed slots) and at maximum `capacity` bytes. Whenever all the 8 slots are tacked or there
-is no more space on the preallocated memory `BufferPool` failover to a `BufferFromSystemMemory`.
+1. **Back Mode**: Default mode where allocations start from the back of the buffer.
+2. **Front Mode**: Used when slots at the back are full but memory can still be reused by moving to
+   the front.
+3. **Alloc Mode**: Falls back to system memory allocation (`BufferFromSystemMemory`) when both back
+   and front sections are full, providing additional capacity but with reduced performance.
 
-Usually, a message is decoded then a response is sent then a new message is decoded, etc.
-So `BufferPool` is optimized for use all the slots then check if the first slot has been dropped
-If so use it, then check if the second slot has been dropped, and so on.
-`BufferPool` is also optimized to drop all the slices.
-`BufferPool` is also optimized to drop the last slice.
+`BufferPool` can only be fragmented between the front and back and between back and end.
```

Comment:
What does this mean? I'm having some problems in understanding how this back, front and end works tbh.

Reply:
Can you expand more?

Reply:
You can only add data at the beginning or end of the buffer array. You cannot add data in the middle.

Like if our buffer is 6 bytes:
`[0, 0, 0, 0, 0, 0]`

You can add data like:
`[0, 0, 0, 0, 0, 1]` 
or:
`[1, 0, 0, 0, 0, 0]` 

but NOT:
`[0, 0, 0, 1, 0, 0]`